### PR TITLE
fix #19130

### LINKED ISSUE #19130: 2.0.44 breaks DbSession in some cases
This commit 12763a146ebc550963b5d368633539701e7884ec breaks `DbSession` in some cases.

### What steps will reproduce the problem?
Calling  the public function `regenerateID($deleteOldSession = false)` without an existing session will crash.

Source:
```php
public function regenerateID($deleteOldSession = false)
    {
        $oldID = session_id();

        // if no session is started, there is nothing to regenerate
        if (empty($oldID)) {
            return;
        }

        parent::regenerateID(false);
        $newID = session_id();
        // if session id regeneration failed, no need to create/update it.
        if (empty($newID)) {
            Yii::warning('Failed to generate new session ID', __METHOD__);
            return;
        }

        $row = $this->db->useMaster(function() use ($oldID) {
            return (new Query())->from($this->sessionTable)
               ->where(['id' => $oldID])
               ->createCommand($this->db)
               ->queryOne();
        });

        if ($row !== false) {
            if ($deleteOldSession) {
                $this->db->createCommand()
                    ->update($this->sessionTable, ['id' => $newID], ['id' => $oldID])
                    ->execute();
            } else {
                $row['id'] = $newID;
                $this->db->createCommand()
                    ->insert($this->sessionTable, $row)
                    ->execute();
            }
        } else {
            // shouldn't reach here normally
            $this->db->createCommand()
                ->insert($this->sessionTable, $this->composeFields($newID, ''))
                ->execute();
        }
    }
```

It is valid to call `regenerateID()` before a session was ever written to the database. The code above tries to retrieve the existing row from the database, but if that fails it goes to the clause annotated with `// shouldn't reach here normally`.

The issue with this path is that it doesn't go through the `writeSession` implementation, but instead uses its own implementation of just calling `composeFields`.

Inside `writeSession` there is a lot more happening:
```php
            if ($this->writeCallback && !$this->fields) {
                $this->fields = $this->composeFields();
            }
            // ensure data consistency
            if (!isset($this->fields['data'])) {
                $this->fields['data'] = $data;
            } else {
                $_SESSION = $this->fields['data'];
            }
            // ensure 'id' and 'expire' are never affected by [[writeCallback]]
            $this->fields = array_merge($this->fields, [
                'id' => $id,
                'expire' => time() + $this->getTimeout(),
            ]);
            $this->fields = $this->typecastFields($this->fields);
            $this->db->createCommand()->upsert($this->sessionTable, $this->fields)->execute();
```

Specifically special code is executed to ALWAYS set `id` and `expire`.

So we have 1 code path `regenerateId` that doesn't always set `id` and `expire` and another that does.
This makes it so that when `expire` is a column in a database, without a default value, the `regenerateId` path fails.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.44
| PHP version      | 
| Operating system |

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19130

### CODE REVIEW

**Comment by samdark** [comment]
Seems to break on PHP 5.4.

**Comment by SamMousa** [comment]
Probably a syntax error on my side. Will have a look tomorrow unless you want to go ahead and fix it now ;-)

**Comment by samdark** [comment]
Take your time ;)

File: framework/web/DbSession.php [hunk 1]
```diff
@@ -166,8 +168,9 @@ public function regenerateID($deleteOldSession = false)
     public function close()
     {
         if ($this->getIsActive()) {
+
```
Comment (samdark) [inline framework/web/DbSession.php:171]:
```suggestion
```

**Review by samdark — COMMENTED** [review]
Is it possible to unit-test the issue?

**Pull request closed without merging**